### PR TITLE
Add Makefile target to push fat manifest for multi-arch images

### PR DESCRIPTION
**Pull request by nitkon**
Signed-off-by: Nitesh Konkar <niteshkonkar@in.ibm.com>

**What this PR does / why we need it**:
This PR adds support to push fat manifest for multi-arch images. We need this
so that we can seamlessly pull and run the test images on multiple platforms. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

File: test/images/Makefile
Comment (ixdy) [inline test/images/Makefile]: is there a reason to make pushing the manifest a separate make target?

File: test/images/image-util.sh
Comment (ixdy) [inline test/images/image-util.sh]: why not just fold this call into the `push()` function?

**Comment by ixdy** [comment]
Please fill out the template in the PR description comment (https://github.com/kubernetes/kubernetes/pull/63453#issue-186141393) too (i.e. what this PR is doing, why you're doing it, etc).

File: test/images/Makefile
Comment (nitkon) [reply test/images/Makefile]: @ixdy : That was just for separation of concerns within the makefile. However, we can also use all-push target for the same. I have updated my PR accordingly.

File: test/images/image-util.sh
Comment (nitkon) [reply test/images/image-util.sh]: @ixdy : Yes we can.  I have updated the PR.

**Review by mkumatag — APPROVED** [review]
LGTM

**Comment by nitkon** [comment]
@ixdy: I have addressed your comments and filled out the template in the PR description.

**Comment by nitkon** [comment]
@ixdy : Polite ping

**Comment by nitkon** [comment]
@ixdy @luxas Polite Ping :-)

**Comment by dims** [comment]
@ixdy @luxas tentatively marking this for 1.11, please remove milestone if it is not appropriate

/milestone v1.11

**Comment by jberkus** [comment]
This needs a SIG if you want it to progress further.

**Comment by jberkus** [comment]
Also, correct commands:

/kind feature
/priority important-longterm

File: test/images/image-util.sh
Comment (ixdy) [inline test/images/image-util.sh]: neat sed trick: to avoid having to escape the slash, you can actually use (nearly) any character as a separator, not just `/`.

i.e. this could be written
`sed -e 's|[^ ]* *|linux/&|g' -e 's/ /,/g'`

**Review by luxas — APPROVED** [review]
/lgtm

**Comment by luxas** [comment]
@jberkus it's a bit fun as no sig other than possibly sig-release owns this code :thinking:

**Comment by ixdy** [comment]
@luxas we really do need to formalize which sig owns build stuff.

File: test/images/image-util.sh
Comment (ixdy) [reply test/images/image-util.sh]: (in case it's not clear, I'm not blocking this PR on this change, just wanted to give you an FYI in case you didn't know.)

Comment (nitkon) [reply test/images/image-util.sh]: @ixdy : Thanks for the review and the tip!

**Comment by nitkon** [comment]
We need to push the multi-arch images into the gcr repo as well.

**Comment by mkumatag** [comment]
@ixdy can someone please help pushing manifest images?

**Comment by mkumatag** [comment]
> We will need the multi-arch images for conformance testing as well.

@dims I can start working on using manifest images once I have manifest images pushed to k8s test registry, for which I need help from @ixdy or someone who has push permissions. :)

**Comment by dims** [comment]
@k8s-oncall @cjwagner can you please help pushing the images?

**Comment by mkumatag** [comment]
@cjwagner
you need to push all the images which will build images and push manifest for all the images under `test/images`

Run `make all WHAT=<folder_name>` for all the images

```
~/k8s_ws/src/k8s.io/kubernetes/test/images on  master ⌚ 22:56:14
$ find . -name VERSION -print
./ipc-utils/VERSION
./clusterapi-tester/VERSION
./net/VERSION
./test-webserver/VERSION
./redis/VERSION
./mounttest-user/VERSION
./no-snat-test-proxy/VERSION
./mounttest/VERSION
./pets/redis-installer/VERSION
./pets/zookeeper-installer/VERSION
./pets/peer-finder/VERSION
./nautilus/VERSION
./liveness/VERSION
./netexec/VERSION
./serve-hostname/VERSION
./cuda-vector-add/VERSION
./logs-generator/VERSION
./entrypoint-tester/VERSION
./nonewprivs/VERSION
./resource-consumer/controller/VERSION
./resource-consumer/VERSION
./goproxy/VERSION
./iperf/VERSION
./jessie-dnsutils/VERSION
./dnsutils/VERSION
./porter/VERSION
./nettest/VERSION
./hostexec/VERSION
./no-snat-test/VERSION
./fakegitserver/VERSION
./n-way-http/VERSION
./kitten/VERSION
./port-forward-tester/VERSION
```

e.g: `make all WHAT=port-forward-tester` will build and push image, manifest for image `port-forward-tester`

**Comment by cjwagner** [comment]
Done.

**Comment by ixdy** [comment]
I think we're now running into the same issues we encountered in https://github.com/kubernetes/release/pull/516 - `manifest-tool` doesn't support docker credential helpers properly, so we can't push the manifest lists to gcr.io.

**Comment by ixdy** [comment]
I think the way we generally work now is to run `gcloud auth configure-docker` once and then just `docker push` after that.

I think `gcloud auth configure-docker` just adds a `credHelpers` section to `~/.docker/config.json`:
```json
...
  "credHelpers": {
    "us.gcr.io": "gcloud",
    "staging-k8s.gcr.io": "gcloud",
    "asia.gcr.io": "gcloud",
    "gcr.io": "gcloud",
    "k8s.gcr.io": "gcloud",
    "eu.gcr.io": "gcloud"
  },
...
```

**Comment by mkumatag** [comment]
@ixdy seems like manifest-tool does not support `credHelpers` kind of auth, so we may need to use `docker manifest` cli which got introduced in `18.02` version. Will be submitting a patch to replace manifest-tool command with docker cli.

**Comment by luxas** [comment]
Thanks @mkumatag!

**Comment by mkumatag** [comment]
@luxas @dims @ixdy @cjwagner created https://github.com/kubernetes/kubernetes/pull/65046 to use `docker manifest` command, PTAL

**Comment by dims** [comment]
Thanks @mkumatag

**Comment by mkumatag** [comment]
@cjwagner can you try now pushing the images with at least `docker 18.03.1` version installed?

**Comment by nitkon** [comment]
@cjwagner : Any update on the push request ?

**Comment by luxas** [comment]
@mkumatag did you notice that there's a problem with using docker manifest in 18.03?
https://github.com/docker/cli/issues/1135

**Comment by mkumatag** [comment]
@luxas above patch has been backported for [`docker 18.06`](https://github.com/docker/cli/pull/1183) version as well and RC2 is out recently. @cjwagner can you try installing RC2 version and push manifest?

**Comment by cjwagner** [comment]
I'm still trying to get this to work...
I failed to get docker 18.06 to install on my corporate workstation, but was able to get the mac version installed on my laptop. After battling some issues related to the default bash version being too old, missing dev tools, and publishing permission errors, I am now getting stuck at a failed build step that I don't know how to resolve:
```sh
./image-util.sh: line 105: /var/folders/n9/7pd1wch53mjf_s4m4sgy3qgc00f148/T/tmp.IRGu0VYE/manifest-tool: cannot execute binary file: Exec format error
```

**Comment by mkumatag** [comment]
@cjwagner looking at your error message what I can say is you are using very old code which contains `manifest-tool` to push the image which is not in the latest code anymore. Can you freshly clone the `https://github.com/kubernetes/kubernetes` repo or pull latest from master branch and try rerunning the steps..

**Note**: I'm not very confident running this on mac but should work! Don't forget to follow the instructions below to make manifest command working:

```
Enable experimental features:
On Linux, as root, edit /etc/docker/daemon.json to contain {"experimental":true}. Then, as root, run service docker restart to apply the setting.
On Mac, open Docker's Preferences pane, then check the "Enable experimental features" checkbox, and click "Apply and restart" to apply the setting.
```

**Comment by cjwagner** [comment]
I thought I did a fresh clone of k/k (I started with an empty dir), but somehow I got an old commit. My local checkout is up to date now though.
Those instructions weren't quite correct. The `experimental` key in the config.json file needs to be set to `enabled`. Changing the checkbox in the preferences pane didn't seem to make a difference.

I was able to build and push `port-forward-tester`, but when I moved on to some of the other images I ran into more problems with some `D` command in the docker files. I don't actually see a reference to such a command in the Dockerfiles though...
```
colew-macbookpro:images colew$ make all-push WHAT=net
./image-util.sh build net
Building image for net ARCH: amd64...
../image-util.sh bin net
/var/folders/n9/7pd1wch53mjf_s4m4sgy3qgc00f148/T/tmp.ZLgahqz5 ~/go/src/k8s.io/kubernetes/test/images
sed: 1: "Dockerfile": extra characters at the end of D command
make: *** [all-container] Error 1

colew-macbookpro:images colew$ make all-push WHAT=ipc-utils
./image-util.sh build ipc-utils
Building image for ipc-utils ARCH: amd64...
/var/folders/n9/7pd1wch53mjf_s4m4sgy3qgc00f148/T/tmp.0dVdDiDy ~/go/src/k8s.io/kubernetes/test/images
sed: 1: "Dockerfile": extra characters at the end of D command
make: *** [all-container] Error 1

colew-macbookpro:images colew$ make all-push WHAT=clusterapi-tester
./image-util.sh build clusterapi-tester
Building image for clusterapi-tester ARCH: amd64...
../image-util.sh bin clusterapi-tester
/var/folders/n9/7pd1wch53mjf_s4m4sgy3qgc00f148/T/tmp.hZzHzL7n ~/go/src/k8s.io/kubernetes/test/images
sed: 1: "Dockerfile": extra characters at the end of D command
make: *** [all-container] Error 1
```

**Comment by dims** [comment]
@cjwagner that's possibly caused by the version of `sed` in your environment. could you try `brew install gnu-sed --with-default-names`?

(found other similar issues at https://www.google.com/search?q=%22%22Dockerfile%22%3A+extra+characters+at+the+end+of+D+command%22)

**Comment by BenTheElder** [comment]
a number of scripts will autodetect `gsed` and prefer that when available,
it'd be nice to do that here as well, FWIW

**Comment by dims** [comment]
@BenTheElder of course, i was giving a hint to see if it gets past the problem(s) then we can make the changes needed

**Comment by BenTheElder** [comment]
right sorry, that sounds correct to me as well, I just wanted to mention that we should probably do that if we aren't already as a follow-up, then you don't need to do `--with-default-names`.

**Comment by cjwagner** [comment]
I can't update sed. My corporate laptop won't allow some of the install steps to run, namely some `conftest` binary. I can't create an exception for it because it runs with a different identifier (file hash I think) every time...
The linking step also fails, but presumably that is because of the previous failures: 
```
==> make install
Last 15 lines from /Users/colew/Library/Logs/Homebrew/gnu-sed/02.make:
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(c-ctype.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(getprogname.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(lock.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(se-context.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(se-selinux.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(stat-time.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(threadlib.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(unistd.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: lib/libsed.a(wctype-h.o) has no symbols
  CCLD     sed/sed
ld: library not found for -lcP
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [sed/sed] Error 1
make[1]: *** [install-recursive] Error 1
make: *** [install] Error 2

READ THIS: https://docs.brew.sh/Troubleshooting
```

What part of the build is dependent on a specific version of sed?(!)

**Comment by dims** [comment]
one of the 5 here @cjwagner (guessing based on the traces above) - https://github.com/kubernetes/kubernetes/blob/master/test/images/image-util.sh#L67-L109

**Comment by dims** [comment]
@BenTheElder @cjwagner please see https://github.com/kubernetes/kubernetes/pull/66585/ just like other scripts we have we need `gsed` available in PATH for this to work

**Comment by BenTheElder** [comment]
Thanks @dims
Oh and:
> What part of the build is dependent on a specific version of sed?(!)

Mac ships with mostly BSD style tools, the scripts use GNU style tools, which have both major and subtle differences. Doing nontrivial sed portably can be a real pain so kubernetes just targets gnu-sed. Dims' fix should make it possible to just `brew install gnu-sed` then run these

**Comment by dims** [comment]
@cjwagner the specific problem you see is

```
-i '' works with OS X's sed but not GNU sed. -i works with GNU sed, recent versions of NetBSD, OpenBSD sed, but not OS X's sed. -i -e works with both but in the case of FreeBSD sed makes a backup of the original file with -e appended to the file name (and you need to pass no more than one expression to sed).
```

(quote from https://unix.stackexchange.com/questions/13711/differences-between-sed-on-mac-osx-and-other-standard-sed)

**Comment by cjwagner** [comment]
OK, thanks @dims! 
I think the images are building and publishing properly now. 🎉 🤞 
I'll update here when everything has finished.

**Comment by dims** [comment]
Very cool @cjwagner thanks!

**Comment by mkumatag** [comment]
@cjwagner Thanks a ton, few more images are missing, can you push them as well(though I don't have any idea why they are pushing the same image into the different repo and using it!)

```
gcr.io/k8s-authenticated-test/serve-hostname:1.1

**Comment by dims** [comment]
@mkumatag - we need to track down why “goproxy” “jessie-dnsutils” “redis” did not get built/pushed too (i added FIXME's in my PR)

**Comment by luxas** [comment]
Can we move this general discussion into an issue please?

**Comment by dims** [comment]
yes, let's please use https://github.com/kubernetes/kubernetes/issues/66618

**Pull request merged**
